### ISSUE TITLE
[AArch64] Passing a v1f32 arg through memory crashes backend with "Invalid TRUNCATE!" assertion error

### ISSUE THREAD

**Issue by HazyFish** [llvm/llvm-project#58615]
## Description

Passing a `v1f32` arg through memory crashes backend with assertion `VT.isInteger() && Operand.getValueType().isInteger() && "Invalid TRUNCATE!"` failed during `AArch64 Instruction Selection` pass when targeting AArch64.

The problem only occurs when the `v1f32` is passed through memory. In the reproduction below, we use a `v16i64` to occupy all 8 registers used for passing arguments so that `v1f32` has to be passed through memory.

## Minimal Reproduction

https://godbolt.org/z/rz863x8dr

### Code

```ll
define <1 x float> @f(<16 x i64> %0, <1 x float> %1) {
BB:
  ret <1 x float> %1
}
```

### Stack Trace

```console
llc: /home/henry/aflplusplus-isel/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:5372: llvm::SDValue llvm::SelectionDAG::getNode(unsigned int, const llvm::SDLoc &, llvm::EVT, llvm::SDValue, const llvm::SDNodeFlags): Assertion `VT.isInteger() && Operand.getValueType().isInteger() && "Invalid TRUNCATE!"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: ./llvm-project/build-debug/bin/llc -mtriple=aarch64 ./crash-reports/dagisel-aarch64/4.ll
1.	Running pass 'Function Pass Manager' on module './crash-reports/dagisel-aarch64/4.ll'.
2.	Running pass 'AArch64 Instruction Selection' on function '@f'
 #0 0x000000000479068a llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Support/Unix/Signals.inc:569:11
 #1 0x000000000479083b PrintStackTraceSignalHandler(void*) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Support/Unix/Signals.inc:636:1
 #2 0x000000000478ee86 llvm::sys::RunSignalHandlers() /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Support/Signals.cpp:104:5
 #3 0x0000000004790f65 SignalHandler(int) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Support/Unix/Signals.inc:407:1
 #4 0x00007f65e46d9980 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x12980)
 #5 0x00007f65e35c9e87 raise /build/glibc-CVJwZb/glibc-2.27/signal/../sysdeps/unix/sysv/linux/raise.c:51:0
 #6 0x00007f65e35cb7f1 abort /build/glibc-CVJwZb/glibc-2.27/stdlib/abort.c:81:0
 #7 0x00007f65e35bb3fa __assert_fail_base /build/glibc-CVJwZb/glibc-2.27/assert/assert.c:89:0
 #8 0x00007f65e35bb472 (/lib/x86_64-linux-gnu/libc.so.6+0x30472)
 #9 0x00000000044929ce llvm::SelectionDAG::getNode(unsigned int, llvm::SDLoc const&, llvm::EVT, llvm::SDValue, llvm::SDNodeFlags) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:5373:5
#10 0x00000000044789fd llvm::SelectionDAG::getNode(unsigned int, llvm::SDLoc const&, llvm::EVT, llvm::SDValue) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:5054:10
#11 0x00000000044790da llvm::SelectionDAG::getAnyExtOrTrunc(llvm::SDValue, llvm::SDLoc const&, llvm::EVT) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:1410:5
#12 0x000000000444df8d getCopyFromPartsVector(llvm::SelectionDAG&, llvm::SDLoc const&, llvm::SDValue const*, unsigned int, llvm::MVT, llvm::EVT, llvm::Value const*, llvm::Optional<unsigned int>) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:421:16
#13 0x00000000043faddc getCopyFromParts(llvm::SelectionDAG&, llvm::SDLoc const&, llvm::SDValue const*, unsigned int, llvm::MVT, llvm::EVT, llvm::Value const*, llvm::Optional<unsigned int>, llvm::Optional<llvm::ISD::NodeType>) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:171:12
#14 0x00000000044496a2 llvm::SelectionDAGISel::LowerArguments(llvm::Function const&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:10713:29
#15 0x00000000044f2cfa llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:1370:3
#16 0x00000000044f1776 llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:467:3
#17 0x000000000117b625 (anonymous namespace)::AArch64DAGToDAGISel::runOnMachineFunction(llvm::MachineFunction&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:55:5
#18 0x0000000003566e65 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/CodeGen/MachineFunctionPass.cpp:91:8
#19 0x0000000003c3c3a6 llvm::FPPassManager::runOnFunction(llvm::Function&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1430:23
#20 0x0000000003c411d2 llvm::FPPassManager::runOnModule(llvm::Module&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1476:16
#21 0x0000000003c3cc79 (anonymous namespace)::MPPassManager::runOnModule(llvm::Module&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1545:23
#22 0x0000000003c3c7ed llvm::legacy::PassManagerImpl::run(llvm::Module&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:535:16
#23 0x0000000003c414b1 llvm::legacy::PassManager::run(llvm::Module&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1672:3
#24 0x0000000000d3f74c compileModule(char**, llvm::LLVMContext&) /home/henry/aflplusplus-isel/llvm-project/llvm/tools/llc/llc.cpp:737:41
#25 0x0000000000d3daf2 main /home/henry/aflplusplus-isel/llvm-project/llvm/tools/llc/llc.cpp:418:13
#26 0x00007f65e35acc87 __libc_start_main /build/glibc-CVJwZb/glibc-2.27/csu/../csu/libc-start.c:344:0
#27 0x0000000000d3d2fa _start (./llvm-project/build-debug/bin/llc+0xd3d2fa)
```

## Cause

We need to copy parts of a `v2i32` into a `v1f32` vector. However, the following code does not handle cases where both the element type and the vector length are different.

https://github.com/llvm/llvm-project/blob/2098e2f472578f793c565b669500dca454b4a608/llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp#L396-L422

**Comment by HazyFish** [comment]
I'm working on a patch. cc @DataCorrupted

**Comment by HazyFish** [comment]
Candidate patch https://reviews.llvm.org/D136726

**Comment by HazyFish** [comment]
Fixed by ef0d689e8be2ac22b2e2da477217c761354c0ff9
